sql/sql_expression.cc

```diff
@@ -3119,49 +3128,246 @@ bind(SqlBindingScope & context) const
-                    storage.mergeToRowDestructive(result);
-                else v.appendToRow(Path(), result);
+        // Have we already done this prefix?
+        std::vector<int> donePrefixes(allPrefixes.size(), 0);
+
+        struct ClauseInstruction {
+            ClauseInstruction(PathElement el = PathElement(),
+                              int outputNumber = -1,
+                              bool isFirst = false)
+                : el(std::move(el)),
+                  outputNumber(outputNumber),
+                  isFirst(isFirst)
+            {
+            }
+
+            PathElement el;
+            int outputNumber = -1;  ///< Where do we put it?
+            bool isFirst = false;   ///< What do we do with it?
+
+            void addToResult(ExpressionValue row,
+                             StructValue & result) const
+            {
+                if (isFirst) {
+                    result[outputNumber] = std::make_tuple(el, std::move(row));
+                }
+                else {
+                    std::get<1>(result[outputNumber])
+                        = ExpressionValue::superpose({std::get<1>(result[outputNumber]),
+                                                      std::move(row)});
+                }
+            }
+        };
+
+        // For each clause, for each prefix, what do we do with it?
+        struct ClauseInstructions {
+            /// Expression for the clause
+            BoundSqlExpression expr;
+
+            /// Instructions for this particular clause
+            std::vector<ClauseInstruction> clauseInstructions;
+
+            ClauseInstructions(BoundSqlExpression expr,
+                               std::vector<ClauseInstruction> clauseInstructions)
+                : expr(std::move(expr)),
+                  clauseInstructions(std::move(clauseInstructions))
+            {
+            }
+
+            void apply(const SqlRowScope & context,
+                       const VariableFilter & filter,
+                       StructValue & result) const
+            {
+                // 1.  Apply the clause and get its value
+
+                ExpressionValue clauseStorage;
+                const ExpressionValue & v = expr(context, clauseStorage, filter);
+
+                StructValue row;
+                if (&v == &clauseStorage) {
+                    clauseStorage.mergeToRowDestructive(row);
+                }
+                else {
+                    v.appendToRow(Path(), row);
+                }
+
+                //cerr << "clause " << clauses[i]->surface
+                //     << " had result " << jsonEncodeStr(row) << endl;
+
+                // If this throws, the ExpressionValueInfo says that there
+                // are a fixed set of columns, but what returned wasn't
+                // the same (there are more or less of them than expected)
+                if (row.size() != clauseInstructions.size()) {
+                    cerr << "row.size() = " << row.size() << endl;
```

Comment:
If we use the logger in cases like this, it'll most likely be turned off if the error happens, which means you wont have the info to help you understand or reproduce the error...